contracts/DKVDaggerHashimoto.sol

```diff
@@ -195,15 +196,53 @@ contract DecentralizedKVDaggerHashimoto is DecentralizedKV {
 
     /*
      * Run a modified hashimoto hash.
-     * This is eaiser running in evm
+     * This is eaiser running in evm, and it is designed for smaller size
      */
     function _hashimotoKeccak256(
         uint256 startShardId,
         uint256 shardLenBits,
         bytes32 hash0,
         bytes[] memory maskedData
     ) internal view returns (bytes32) {
-        require(maskedData.length == randomChecks, "incorrect PoRA");
+        /* Assumption check */
+        require(maskedData.length == randomChecks, "data vs checks: length mismatch");
+        if (randomChecks != 0) require(maskedData[0].length <= SYSTEM_BLOCK_SIZE, "too large data uploaded");
+
+        uint256 maxKvSize = 1 << maxKvSizeBits;
+        uint256 rows = 1 << (shardEntryBits + shardLenBits);
+
+        for (uint256 i = 0; i < randomChecks; i++) {
+            require(maskedData[i].length == maxKvSize, "invalid proof size");
+            uint256 parent = uint256(hash0) % rows;
+            uint256 kvIdx = parent + (startShardId << shardEntryBits);
+            bytes memory data = maskedData[i];
+            require(systemContract.checkDaggerData(kvIdx, kvMap[idxMap[kvIdx]].hash, data),
+                    "invalid access proof");
+
+            assembly {
+                mstore(data, hash0)
+                hash0 := keccak256(data, add(maxKvSize, 0x20))
+                mstore(data, maxKvSize)
+            }
+        }
+        return hash0;
+    }
+
+    /*
+     * Run a modified hashimoto hash,
+     * with Merkle inclusion proofs
+     */
+    function _hashimotoMerkleProof(
+        uint256 startShardId,
+        uint256 shardLenBits,
+        bytes32 hash0,
+        uint256[] memory checkIdList,
+        bytes32[][] memory proofsDim2,
+        bytes[] memory maskedData
+    ) internal view returns (bytes32) {
+        require(maskedData.length == randomChecks, "data vs checks: length mismatch");
+        require(proofsDim2.length == randomChecks, "proofs vs checks: length mismatch");
+        require(checkIdList.length == randomChecks, "proofs idx vs checks: length mismatch");
```

Comment:
I don't think we should pass checkIdList - instead, `rows` will be the number of 4KB chunks in a shard, and we will find `chunkIdx` instead of `kvIdx`, and we will also know the position of the chunk in the large value of the KV.